### DISCUSSION TITLE
Does next-shared-cache support  stale-while-revalidate?

### DISCUSSION THREAD

**Discussion by alex-guoba** [caching-tools/next-shared-cache#489] [Q&A] [answered]
Hi, In my project when the key/value has expired, I noticed the next request rendring was really slow. because next.js will fetch data from backend server and write to redis by next-share-cache.

Next data cache has provided a `[stale-while-revalidate](https://nextjs.org/docs/app/building-your-application/caching#time-based-revalidation)` behaivor.  How to implement it in cache-handler ?

**Comment by better-salmon** [comment, marked as answer]
Hey, @alex-guoba.

You can use the `estimateExpireAge` callback to estimate the expiration age of stale data. By default, the stale age is the same as the expiration age, which can cause long loading times when the data becomes stale. However, you can customize the expiration age or calculate it based on the stale age to improve loading times.

```js
CacheHandler.onCreation(async () => {
  const handler = {
    // ...
  };
 
  return {
    handlers: [handler],
    ttl: { estimateExpireAge: (staleAge) => staleAge * 10 }, // 10 time more than stale age
    // or set fixed expire age in seconds by doing like this
    // ttl: { estimateExpireAge: () => 60 * 60 * 24 * 365 }, // 1 year
  };
});
```

Here is [the documentation for the `estimateExpireAge`](https://caching-tools.github.io/next-shared-cache/api-reference/ttl-parameters#estimateexpireage).

I hope you find this helpful!

**Reply by alex-guoba** [reply]
Thanks, it's really works.    How does it works ?  I means next.js framework only exposed 3 basic [API](https://nextjs.org/docs/app/api-reference/next-config-js/incrementalCacheHandlerPath) to developer，what did cache-handler do to tell next.js to return first and then update cache asynchronously.

**Reply by better-salmon** [reply]
Actually, it does nothing special. If CacheHandler returns a stale value, Next.js revalidates it in the background. Native cache works the same. However, CacheHandler enriches the cache value with some [metadata about its expiration parameters](https://caching-tools.github.io/next-shared-cache/api-reference/lifespan-parameters). Handlers can use this metadata to set the TTL option in a cache store, which can expire values on the cache storage side.

It would be best to find a balance between the long expiration age and the available memory in your cache store. A longer expiration age means more memory consumption.

**Reply by alex-guoba** [reply]
Good suggestion. I will recaculate my expiration time according to my redis storagize size.  Thanks again.

**Comment by FanManutd** [comment]
> It would be best to find a balance between the long expiration age and the available memory in your cache store. A longer expiration age means more memory consumption.

@alex-guoba 

Hi! I have a clarifying question. If you set the expiration age to be greater than the stale age, then when Next revalidates a cache entry, will it overwrite the current entry or create a new one?

If Next overwrites the current entry, then there should be no problems with memory, even if you set the expiration date to a year. Because Next will overwrite entries in the cache, which is as close as possible to the stale while revalidate. Or I'm wrong?

**Reply by alex-guoba** [reply]
when Next revalidates a cache entry, will it overwrite the current entry or create a new one?
——————
Next.js will overwrite the cache with the same key.  expiration age was saved in the value structure.

This will still be a problem because redis will not expirate the key/value because the expiration age not reached, it still consume the memory of redis-server.  Memory is expensive and provide high performance then file. Consider the scenario if you have millions of data in backend server(like database), but limited capacity in redis.
It wouldn't  be a problem if your redis-server have enough capacity can cache all your data.

**Reply by FanManutd** [reply]
I got it, thank you!
